crates/adapters/src/static_compile/serialize_with_context.rs

```diff
@@ -0,0 +1,489 @@
+serialize_without_context!(u8);
+serialize_without_context!(i16);
+serialize_without_context!(u16);
+serialize_without_context!(i32);
+serialize_without_context!(u32);
+serialize_without_context!(i64);
+serialize_without_context!(u64);
+serialize_without_context!(i128);
+serialize_without_context!(u128);
+serialize_without_context!(usize);
+serialize_without_context!(isize);
+serialize_without_context!(String);
+serialize_without_context!(char);
+serialize_without_context!(F32);
+serialize_without_context!(F64);
+serialize_without_context!(Decimal);
+
+/// Used to pass the context to nested structures during serialization.
+// This is only public because it's used in a macro; it's not suppposed
+// to be user-visible otherwise.
+#[doc(hidden)]
+pub struct SerializationContext<'se, C, T> {
+    context: &'se C,
+    value: &'se T,
+    phantom: PhantomData<T>,
+}
+
+impl<'se, C, T> SerializationContext<'se, C, T> {
+    pub fn new(context: &'se C, value: &'se T) -> Self {
+        Self {
+            context,
+            value,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'se, C, T> Serialize for SerializationContext<'se, C, T>
+where
+    T: SerializeWithContext<C>,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.value.serialize_with_context(serializer, self.context)
+    }
+}
+
+impl<C, T> SerializeWithContext<C> for Vec<T>
+where
+    T: SerializeWithContext<C>,
+{
+    fn serialize_with_context<S>(&self, serializer: S, context: &C) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.len()))?;
+        for element in self {
+            seq.serialize_element(&SerializationContext::new(context, element))?;
+        }
+
+        seq.end()
+    }
+}
+
+impl<C, T> SerializeWithContext<C> for &T
+where
+    T: SerializeWithContext<C>,
+{
+    fn serialize_with_context<S>(&self, serializer: S, context: &C) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        T::serialize_with_context(*self, serializer, context)
+    }
+}
+
+macro_rules! serialize_tuple {
+    ([$num_fields:expr]($(($arg_name:ident, $arg_type:tt)),*)) => {
+        #[allow(unused_variables)]
+        #[allow(dead_code)]
+        #[allow(unused_mut)]
+        impl<C, $($arg_type),*> SerializeWithContext<C> for ($($arg_type),*)
+        where
+            $($arg_type: SerializeWithContext<C>),*
+        {
+            fn serialize_with_context<S>(&self, serializer: S, context: &C) -> Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                let ($($arg_name),*) = self;
+                let mut tuple = serializer.serialize_tuple($num_fields)?;
+                $(tuple.serialize_element(&SerializationContext::new(context, $arg_name))?;)*
+                tuple.end()
+            }
+        }
+    }
+}
+
+serialize_tuple!([0]());
```

Comment:
can you write a macro to generate these?

Reply:
heh, be my guest :)